include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -1324,6 +1324,292 @@ namespace Differentiation
     }; // class ADHelperEnergyFunctional
 
 
+
+    /**
+     * A helper class that facilitates the evaluation and automated
+     * linearization of a vector of functions that represents a residual vector
+     * (as computed from some corresponding local degree-of-freedom values).
+     * This class would typically be used to compute the linearization of a
+     * residual vector defined on the level of a cell, or for local
+     * nonlinear equations.
+     *
+     * An example of its usage in the case of a residual linearization
+     * might be as follows (in this case we'll compute the
+     * linearization of a finite-strain magneto-elastic solid from the residual,
+     * as constructed from the Piola-Kirchoff stress and magnetic induction
+     * assuming the magnetic scalar potential formulation):
+     *
+     * @code
+     *   // Existing data structures:
+     *   Vector<double> solution (...); // Or another vector type
+     *   std::vector<types::global_dof_index> local_dof_indices (...);
+     *   const FEValuesExtractors::Vector u_fe (...);
+     *   const FEValuesExtractors::Scalar msp_fe (...);
+     *   const unsigned int u_block (...);
+     *   const unsigned int msp_block (...);
+     *   FEValues<dim> fe_values (...);
+     *   const unsigned int n_q_points (...);
+     *   FullMatrix<double> cell_matrix (...);
+     *   Vector<double> cell_rhs (...);
+     *
+     *   // Assembly loop:
+     *   for (auto cell & : ...)
+     *   {
+     *     cell->get_dof_indices(local_dof_indices);
+     *     const unsigned int n_independent_variables
+     *       = local_dof_indices.size();
+     *     const unsigned int n_dependent_variables
+     *       = local_dof_indices.size();
+     *
+     *     // Create some aliases for the AD helper.
+     *     // In this example, we strictly assume that we're using tapeless
+     *     // AD types, and so the AD_typecode used in the template argument
+     *     // must refer to one of these types. See the example for the
+     *     // ADHelperEnergyFunctional for details on how to extend
+     *     // support to taped AD numbers.
+     *     using ADHelper = AD::ADHelperEnergyFunctional<...>;
+     *     using ADNumberType = typename ADHelper::ad_type;
+     *
+     *     // Create and initialize an instance of the helper class.
+     *     ADHelper ad_helper(n_independent_variables,n_dependent_variables);
+     *
+     *     // Initialize the local data structures for assembly.
+     *     // This is also taken care of by the ADHelper, so this step could
+     *     // be skipped.
+     *     cell_rhs.reinit(n_dependent_variables);
+     *     cell_matrix.reinit(n_independent_variables,n_dependent_variables);
```

Comment:
In most codes, we simply write
```
  cell_rhs = 0;
  cell_matrix = 0;
```
because these matrices are already correctly sized. Couldn't we do the same here?

Reply:
I'm trying to show how these should be sized in the event that a `hp::DofHandler` is used. I think its useful (and its consistent with the example given for `ADHelperEnergyFunctional`) so I prefer for this to stay as is.

Reply:
Oh, good point with the hp case!